pkg/schema/type.go

```diff
@@ -66,101 +72,107 @@ func (m ScalarType) GetValueType() yamlmeta.Type {
-	violationErrorMessage := prependErrorMessage + " was type %T when %T was expected"
-
+func (a *ArrayType) CheckType(node yamlmeta.TypeWithValues) (chk yamlmeta.TypeCheck) {
 	_, ok := node.(*yamlmeta.Array)
 	if !ok {
-		scalar, ok := node.(*yamlmeta.Scalar)
-		if ok {
-			chk.Violations = append(chk.Violations, fmt.Sprintf(violationErrorMessage, scalar.Value, &yamlmeta.Array{}))
-		} else {
-			chk.Violations = append(chk.Violations, fmt.Sprintf(violationErrorMessage, node, &yamlmeta.Array{}))
-		}
+		chk.Violations = append(chk.Violations,
+			NewTypeError(node.ValueTypeAsString(), typeToString(a), node.GetPosition(), a.Position))
 	}
 	return
 }
-func (a ArrayItemType) CheckType(node yamlmeta.TypeWithValues, prependErrorMessage string) (chk yamlmeta.TypeCheck) {
-	violationErrorMessage := prependErrorMessage + " was type %T when %T was expected"
 
+func (a *ArrayItemType) CheckType(node yamlmeta.TypeWithValues) (chk yamlmeta.TypeCheck) {
 	_, ok := node.(*yamlmeta.ArrayItem)
 	if !ok {
-		chk.Violations = append(chk.Violations, fmt.Sprintf(violationErrorMessage, node, &yamlmeta.ArrayItem{}))
-		return
+		// An Array must've yielded a non-ArrayItem which is not valid YAML
+		log.Panicf("ArrayItem type check was called on a non-ArrayItem: %#v", node)
 	}
 	return
 }
-func (m ScalarType) CheckType(node yamlmeta.TypeWithValues, prependErrorMessage string) (chk yamlmeta.TypeCheck) {
-	violationErrorMessage := prependErrorMessage + " was type %T when %T was expected"
-
+func (m *ScalarType) CheckType(node yamlmeta.TypeWithValues) (chk yamlmeta.TypeCheck) {
 	value := node.GetValues()[0]
 	switch itemValueType := value.(type) {
 	case string:
 		if _, ok := m.Type.(string); !ok {
-			violation := fmt.Sprintf(violationErrorMessage, itemValueType, m.Type)
-			chk.Violations = append(chk.Violations, violation)
+			chk.Violations = append(chk.Violations,
+				NewTypeError(node.ValueTypeAsString(), typeToString(m.Type), node.GetPosition(), m.Position))
 		}
 	case int:
 		if _, ok := m.Type.(int); !ok {
-			violation := fmt.Sprintf(violationErrorMessage, itemValueType, m.Type)
-			chk.Violations = append(chk.Violations, violation)
+			chk.Violations = append(chk.Violations,
+				NewTypeError(node.ValueTypeAsString(), typeToString(m.Type), node.GetPosition(), m.Position))
 		}
 	case bool:
 		if _, ok := m.Type.(bool); !ok {
-			violation := fmt.Sprintf(violationErrorMessage, itemValueType, m.Type)
-			chk.Violations = append(chk.Violations, violation)
+			chk.Violations = append(chk.Violations,
+				NewTypeError(node.ValueTypeAsString(), typeToString(m.Type), node.GetPosition(), m.Position))
 		}
 	default:
-		violation := fmt.Sprintf(violationErrorMessage, itemValueType, m.Type)
-		chk.Violations = append(chk.Violations, violation)
+		chk.Violations = append(chk.Violations,
+			NewTypeError(typeToString(itemValueType), typeToString(m.Type), node.GetPosition(), m.Position))
 	}
 	return
 }
 
+func typeToString(value interface{}) string {
```

Comment:
this would be nice to invert onto yamlmeta.Typeable. e.g. Typeable could have AsDebugString() or something like this.

Reply:
Which "this" are you referring to?

Reply:
this function.

Reply:
TypeWithValues now has a "ValueTypeAsString" method which is implemented by each of the nodes. The function is still around (though we can stop exporting it) for the "item" nodes and scalars.